### PR TITLE
3004: E2e tests visualisation filter

### PR DESCRIPTION
**Pull request by kael89**
### Issue #:
https://github.com/beyondessential/tupaia-backlog/issues/3004

That should be the last core test-related change in the current e2e sprint.

An example of using `config.json` for running e2e tests for https://github.com/beyondessential/tupaia-backlog/issues/797:
```json
{
  "dashboardReports": {
    "allowEmptyResponse": false,
    "snapshotTypes": ["responseBody", "html"],
    "urlFiles": [
      "cypress/config/dashboardReportUrls/covidau.json",
      "cypress/config/dashboardReportUrls/default.json"
    ],
    "filter": {
      "dataBuilder": ["sumAllPerMetric", "sumLatestPerMetric"],
      "project": "covidau"
    }
  },
  "mapOverlays": {
    "allowEmptyResponse": false,
    "snapshotTypes": ["responseBody"],
    "urlGenerationOptions": {
      "project": "covidau"
    }
  }
}

```

### CODE REVIEW

File: packages/web-frontend/cypress/scripts/generateConfig/VisualisationFilter.js
Comment (kael89) [inline packages/web-frontend/cypress/scripts/generateConfig/VisualisationFilter.js:47]: I used `or` logic here. I guess that `and` would be also useful, but perhaps we could change this when required?

File: packages/web-frontend/cypress/scripts/generateConfig/generateOverlayConfig.js
Comment (kael89) [inline packages/web-frontend/cypress/scripts/generateConfig/generateOverlayConfig.js:63]: Some unhappy connotation with `objectification` here... But it sounds cool in code (?)

File: packages/web-frontend/cypress/scripts/generateConfig/VisualisationFilter.js
Comment (kael89) [reply packages/web-frontend/cypress/scripts/generateConfig/VisualisationFilter.js:47]: Actually, it wouldn't be too hard to add support for `and` - we'd just need to think a bit of how to express it in config.

I know there is existing precedent of using different syntax than the following, but one option would be:

```jsonc
   {
     // `or` logic 
     "filter": { "dataBuilder": "sum", "project": "covidau" },
     // `and` logic
     "filter": [{ "dataBuilder": "sum" }, { "project": "covidau" }]
   }
   ```
   
I'm keen on suggestions - and should probably look into existing conventions

File: packages/web-frontend/cypress/scripts/generateConfig/generateOverlayConfig.js
Comment (rohan-bes) [reply packages/web-frontend/cypress/scripts/generateConfig/generateOverlayConfig.js:63]: That's just what they thought when implementing master and slave servers, and look how that's ended up 😝

If is a nice parallel to `stringifyUrl`... `JSON` library uses `JSON.parse(string)` and `JSON.stringify(object)`... how does `parseUrl(string)` sound?

Comment (kael89) [reply packages/web-frontend/cypress/scripts/generateConfig/generateOverlayConfig.js:63]: Sounds good! Can change that

File: packages/web-frontend/cypress/scripts/generateConfig/configSchema.js
Comment (kael89) [inline packages/web-frontend/cypress/scripts/generateConfig/configSchema.js:20]: @rohan-bes we have discussed in the past about config validation... You may want to have a look at the validation approach taken here (https://github.com/beyondessential/tupaia/pull/2694/commits/06a34f8403ec20c78f40d1176ef67ad815048a08)

I used [yup](https://www.npmjs.com/package/yup), which is a cute validation library also used in Tamanu. Other options considered (briefly):
* [joi](https://github.com/sideway/joi), which seems like a more powerful version of `yup`
* [ajv](https://www.npmjs.com/package/ajv), which is a [JSON schema](https://json-schema.org/) implementation. I really like how VSCode uses JSON schema to enable config validation and autocompletion (eg when editing `eslint` or `jest` config).

For our purposes here and given that we already use `yup` in Tamanu, I decided to go with yup (had a brief discussion with @edmofro about this too).

You may be interested in the above for introducing config validation in `report-server`, and I sense we may want to leverage this in the Admin Panel Viz Builder. Apart from validation, `yup` also handles casting (filling in empty fields, converting between types etc). Let me know if you need more info!

File: packages/web-frontend/cypress/e2e/mapOverlays.spec.js
Comment (kael89) [inline packages/web-frontend/cypress/e2e/mapOverlays.spec.js:21]: It would be good to get a second opinion on this... Here we include all response fields in the snapshots. However, many of them are metadata, and some of them are not currently displayed in the front end. So, even if there are any changes in them, users would not notice.

A `measureData` response example:

```json
{
  "chartType": "bar",
  "dashboardGroupId": "88",
  "data": [
    { "name": "ACT", "value": 124 },
    { "name": "NSW", "value": 5729 }
  ],
  "name": "COVID-19 Total Confirmed Cases by State",
  "organisationUnitCode": "AU",
  "period": {
    "earliestAvailable": "20200125",
    "latestAvailable": "20210626",
    "requested": "201701;201702..."
  },
  "showPeriodRange": "all",
  "type": "chart",
  "viewId": "COVID_Total_Cases_By_State"
}
```

Our main focus is in `data`, but we also show other fields in the front end (eg `period.latestAvailable`). So I am not sure if using the whole response is the best way forward:

**Pros**
* No need to worry about updating e2e tests if fields change/are exposed in the front end
* Metadata don't change often, so including them should not frequently result in false negatives (?)
* We currently don't use component (DOM) testing for overlays, so including metadata such as `name` in the snapshots can be kind of a substitute. This does not assert that the metadata are actually rendered correctly, but the same is true for `data` too

**Cons**
* Bigger snapshots
* Tests may fail for unimportant reasons (eg a field that was deprecated)

I was thinking that we can use the "safest" approach for now, and revisit this after we get some feedback from actual usage. Another approach would be adding config that controls this behaviour (eg `includeMetadata`)

File: packages/web-frontend/cypress/scripts/generateConfig/generateOverlayConfig.js
Comment (edmofro) [inline packages/web-frontend/cypress/scripts/generateConfig/generateOverlayConfig.js:29]: Clever de-duping algorithm!

**Review by edmofro — APPROVED** [review]
Great feature to add! Also crafted in a very clever way. I do think in some cases it could be verging on too clever, where added code complexity results in diminishing returns, or abstraction to achieve genericism makes things harder to understand than repeating a bit of extra code, but given it's all built nicely already and I'm sure works a treat, very happy to approve it! 

Minor things inline, but most comments don't require any action. A couple of superficial things can be changed post-review without another approval.

Nice work @kael89!

File: packages/web-frontend/cypress/scripts/generateConfig/configSchema.js
Comment (kael89) [reply packages/web-frontend/cypress/scripts/generateConfig/configSchema.js:14]: I actually wrapped them to make the validation more readable! Without it, the schema becomes harder to parse, in my opinion...

Will post an example of inline statements, then we can decide together!

Comment (kael89) [reply packages/web-frontend/cypress/scripts/generateConfig/configSchema.js:14]: About using `mixed`: this is just because the `oneOf` method is a child of `mixed`, not `yup` - see https://github.com/jquense/yup#mixedoneofarrayofvalues-arrayany-message-string--function-schema-alias-equals

Comment (kael89) [reply packages/web-frontend/cypress/scripts/generateConfig/configSchema.js:14]:
```js
/**
 * Tupaia
 * Copyright (c) 2017 - 2020 Beyond Essential Systems Pty Ltd
 */

import { yup, yupUtils } from '@tupaia/utils';

import { SNAPSHOT_TYPES } from '../../constants';

const { oneOfType, oneOrArrayOf } = yupUtils;

export const configSchema = yup.object({
  dashboardReports: yup
    .object({
      allowEmptyResponse: yup.boolean(),
      filter: yup
        .object({
          code: oneOrArrayOf(yup.string()),
          project: oneOrArrayOf(yup.string()),
          orgUnit: oneOrArrayOf(yup.string()),
          dashboardGroup: oneOrArrayOf(yup.string()),
          startDate: oneOrArrayOf(yup.string()),
          endDate: oneOrArrayOf(yup.string()),
          dataBuilder: oneOrArrayOf(yup.string()),
        })
        .noUnknown(),
      snapshotTypes: yup
        .array()
        .of(yup.mixed().oneOf([SNAPSHOT_TYPES.RESPONSE_DATA, SNAPSHOT_TYPES.HTML]))
        .min(1)
        .required(),
      urlFiles: yup.array().of(yup.string()),
      urls: yup.array().of(oneOfType([yup.string(), yup.object()])),
    })
    .noUnknown(),
  mapOverlays: yup.object({
    allowEmptyResponse: yup.boolean(),
    filter: yup
      .object({
        id: oneOrArrayOf(yup.string()),
        project: oneOrArrayOf(yup.string()),
        orgUnit: oneOrArrayOf(yup.string()),
        measureBuilder: oneOrArrayOf(yup.string()),
      })
      .noUnknown(),
    snapshotTypes: yup
      .array()
      .of(yup.mixed().oneOf([SNAPSHOT_TYPES.RESPONSE_DATA]))
      .min(1)
      .required(),
    urlFiles: yup.array().of(yup.string()),
    urlGenerationOptions: yup
      .object({
        id: oneOrArrayOf(yup.string()),
        project: oneOrArrayOf(yup.string()),
        orgUnit: oneOrArrayOf(yup.string()),
      })
      .noUnknown(),
    urls: yup.array().of(oneOfType([yup.string(), yup.object()])),
  }),
});
```

Comment (kael89) [reply packages/web-frontend/cypress/scripts/generateConfig/configSchema.js:14]: Perhaps `strings`, `stringOrStrings` and `stringsOrObjects` are not worth it, but I feel that `arrayOf` is - we could add the latter in `@tupaia/utils/src/validation/yup.js` - thoughts @edmofro ?

File: packages/web-frontend/cypress/scripts/generateConfig/generateReportConfig.js
Comment (kael89) [reply packages/web-frontend/cypress/scripts/generateConfig/generateReportConfig.js:54]: You mean with [generateOverlayConfig.js](https://github.com/beyondessential/tupaia/blob/391b8a6f6e19f354b8e41151920228f9a025f054/packages/web-frontend/cypress/scripts/generateConfig/generateOverlayConfig.js)? It seems a bit contradicting with what I did in the filter builders, but I felt that keeping things explicit and simple was worth it!

There's one more reason to do so now that reports and overlays use different identifier keys (`code`/`id`)

Comment (edmofro) [reply packages/web-frontend/cypress/scripts/generateConfig/generateReportConfig.js:54]: Sorry I wasn't clear - I meant shared with web-frontend, but now I realise that the tests should be independent of that!

File: packages/utils/src/__tests__/array.test.js
Comment (kael89) [reply packages/utils/src/__tests__/array.test.js:42]: Good idea!

File: packages/web-frontend/cypress/scripts/generateConfig/configSchema.js
Comment (edmofro) [reply packages/web-frontend/cypress/scripts/generateConfig/configSchema.js:14]: Good thoughts @kael89, and that compromise of wrapping up to a point sounds good. 

I would just like to get @mcccclean's thoughts on it though because this could set a precedent. 

@mcccclean in Tamanu what is your approach? Do you wrap up `yup` type chains into functions?

Pros:
- More readable and concise
Cons:
- Need to understand one extra thing when reading inline
- Potentially less consistent with Tamanu (and also google)?

File: packages/web-frontend/cypress/scripts/generateConfig/VisualisationFilter.js
Comment (kael89) [reply packages/web-frontend/cypress/scripts/generateConfig/VisualisationFilter.js:133]: Yeah things got a bit simpler with https://github.com/beyondessential/tupaia/pull/2781 - we decided that the abstract getters may be an overkill in the JS world, so hopefully now the shared code is easier to read!

I am open to a different approach though - in fact, lately I've been trying to focus on simplicity rather than abstraction. I try to see this from a practical standpoint: what makes code really easier to maintain? Sometimes, having to update code in two different places is quicker and easier than understanding a single piece of abstraction that is hard to understand. The only drawback with the simpler approach is that it may be harder for people to track all the places where an update should take place, but that's what PR reviews/manual/automated testing is for.

**Comment by kael89** [comment]
@edmofro merging this to keep us going, but happy to apply further improvements if we decide to do so in any of the open threads!

File: packages/web-frontend/cypress/scripts/generateConfig/VisualisationFilter.js
Comment (edmofro) [reply packages/web-frontend/cypress/scripts/generateConfig/VisualisationFilter.js:136]: I've come around big time in the same way! You put that very well btw :-)

File: packages/web-frontend/cypress/scripts/generateConfig/configSchema.js
Comment (mcccclean) [reply packages/web-frontend/cypress/scripts/generateConfig/configSchema.js:14]: We actually had this same discussion in Tamanu recently! (although re a much smaller change, so I didn't give it as much thought). The outcome of that was, basically: yes, custom yup validators _should_ be wrapped in functions. The primary thought behind it was just that `yup.string()` and `myCustomYupValidator()` really should have a consistent signature, even if the custom validator doesn't accept any options.

(I guess it's also the case that if it _does_ end up taking options in the future, it doesn't require a refactor!)

Comment (edmofro) [reply packages/web-frontend/cypress/scripts/generateConfig/configSchema.js:14]: Thanks for the input @mcccclean, good to know what Tamanu's doing! Let's stick with that strategy @kael89 (which I think you have here :-))

**Pull request merged**
